lib/usdUfe/ufe/UsdAttributeHolder.cpp

```diff
@@ -38,6 +38,146 @@ namespace {
+        std::replace(group.begin(), group.end(), '|', ':');
+        attr.SetDisplayGroup(group);
+        return true;
+    } else if (key == UsdUfe::MetadataTokens->UIName) {
+        attr.SetDisplayName(value.get<std::string>());
+        return true;
+    }
+
+    return false;
+}
+
+bool hasUsdNativeMetadata(const PXR_NS::UsdAttribute& attr, const std::string& key)
+{
+    PXR_NAMESPACE_USING_DIRECTIVE
+    if (!UsdShadeInput::IsInput(attr) && !UsdShadeOutput::IsOutput(attr)) {
+        // Only allow for shader ports at this time. Could potentially expand
+        // to dynamic attributes if requested. Editing static attributes is
+        // not recommended.
+        return false;
+    }
+    if (key == UsdUfe::MetadataTokens->UIDoc) {
+        return attr.HasAuthoredDocumentation();
+    } else if (key == UsdUfe::MetadataTokens->UIEnumLabels) {
+        return attr.HasMetadata(SdfFieldKeys->AllowedTokens);
+    } else if (key == UsdUfe::MetadataTokens->UIFolder) {
+        return attr.HasAuthoredDisplayGroup();
+    } else if (key == UsdUfe::MetadataTokens->UIName) {
+        return attr.HasAuthoredDisplayName();
+    }
+    return false;
+}
+
+Ufe::Value getUsdNativeMetadata(const PXR_NS::UsdAttribute& attr, const std::string& key)
+{
+    PXR_NAMESPACE_USING_DIRECTIVE
+    if (!hasUsdNativeMetadata(attr, key)) {
+        return {};
+    }
+    Ufe::Value retVal;
+    if (key == UsdUfe::MetadataTokens->UIDoc) {
+        retVal = Ufe::Value(attr.GetDocumentation());
+    } else if (key == UsdUfe::MetadataTokens->UIEnumLabels) {
+        VtTokenArray allowedTokens;
+        if (attr.GetMetadata(SdfFieldKeys->AllowedTokens, &allowedTokens)) {
+            std::string enumstrings;
+            for (auto const& tok : allowedTokens) {
+                if (!enumstrings.empty()) {
+                    enumstrings += ", ";
+                }
+                enumstrings += tok.GetString();
+            }
+            retVal = Ufe::Value(enumstrings);
+        }
+    } else if (key == UsdUfe::MetadataTokens->UIFolder) {
+        std::string uifolder = attr.GetDisplayGroup();
+        // Translate ':' to '|'.
+        // This is to transform nested group separators that differ between platform.
+        // MaterialX uses "/". See "NodeDef Parameter Interface" in the spec.
+        // USD uses ":". See documentation for "SetDisplayGroup".
+        // UFE uses "|". Undocumented, but used in LookdevX.
+        // All three agree that the topmost group is on the left.
+        std::replace(uifolder.begin(), uifolder.end(), ':', '|');
+        retVal = Ufe::Value(uifolder);
+    } else if (key == UsdUfe::MetadataTokens->UIName) {
+        retVal = Ufe::Value(attr.GetDisplayName());
+    }
+    return retVal;
+}
+
+// Returns true if handled, and set cleared on success
+bool clearUsdNativeMetadata(const PXR_NS::UsdAttribute& attr, const std::string& key, bool& cleared)
+{
+    PXR_NAMESPACE_USING_DIRECTIVE
+    cleared = false;
```

Comment:
Not sure why there is this cleared variable, it gets set but never used. Might be a remnant of older code? Or where all the "return true"meant to be "return cleared"?

Reply:
Sadly I have two states to return: handled and cleared. One is done by value, the other one by reference. Could also use a tri-state enum (unhandled, cleared, nothing to clear) if it makes things simpler.